### PR TITLE
Add mypy + some other misc. changes

### LINKED ISSUE #53: Make code more idiomatic, clean, and better enforced
I was working on a separate issue and I started cleaning up the code. I realized that the code needs a little more love to be in tip-top shape and it would be inappropriate to combine a PR that both adds a feature and does a lot of cleanup, so I believe this should be a separate issue.

## Example issues with existing code style

- Incomplete typing, most notably in `hooks.py`.
- Mypy should be part of pre-commit
- pre-commit is not run as part of CI, but should be
- When I substitute some types from the docstrings into Python types, I get a couple type errors in Mypy. Two examples:
   - `FivetranHook.set_schedule_type` says it returns a string, but it clearly returns a dict, and mypy makes this clear. (Upon closer inspection, the return type doesn't actually matter because the return value is never used.)
   - `FivetranHook._do_api_call`'s `json` arg says it takes a dict, but it clearly takes a string, or more specifically a string representation of a JSON.
- Some unnecessary calls to the API. For example, `FivetranHook.prep_connector` makes 3 total API calls when it really only needs to make 2.
- Some un-idiomatic Airflow patterns. For example, `_get_hook()` instead of `hook` as a `cached_property`.
- `FivetranHook._do_api_call` has some unnecessary weird behavior. Notably, it has a parameter called `json` that maps to either the `data=json` or `params=json` kwargs in the `requests.request()` call. This seems less desirable than just passing `data=data`, `params=params`, or `json=json`. Specifically, the `json` kwarg is confusing from the perspective of the `requests` and `aiohttp` APIs: in those APIs, `json=` expects a dict and `data=` expects a string, but in the FivetranHook API, `json=` is a string that gets mapped to `data=`.

## API Breakage?

It's unclear how much API breakage would be tolerated here. We can discuss in a PR, but TLDR I believe for a **minor version bump** that:

- absolutely _zero_ breakage should be tolerated for the `FivetranSensor` and `FivetranOperator`.
- very minor breakage can be tolerated for `FivetranHook`. Specifically for private methods, and methods where the return type is never actually used by anything in `FivetranSensor` or `FivetranOperator`:
  - `_do_api_call` and `_do_api_call_async` should be changed to look and feel like the `requests.request` and `aiohttp.ClientSession.request` APIs.
     - That said, an error message should clearly state to other users what the issue is by checking whether the first arg to these calls is a `tuple`. I imagine there are some people who are poking under the hood of this code and are calling `_do_api_call()` directly.
  - `FivetranHook.check_connector` should return the `connector_details` dict object. (So that `prep_connector` can avoid making an unnecessary 3rd API call.)
  - `FivetranHookprep_connector` should return either `None` or `dict[str, Any] | None`. (Note that the returned value is never used anywhere in the code).

### PR DESCRIPTION
**Pull request by dwreeves**
Resolves #53

# Changes

## Docstring clarification for `FivetranSensor`

I added this because I thought I was working on #49 when I first started coding.

In any event, I added language clarifying how `FivetranSensor` works more specifically than what was there.

## Type-checking

A lot of `hooks.py` was un-annotated, so I made sure to get the whole thing annotated.

I then added Mypy to the pre-commit. I also added `mypy` to the CI.

## `_do_api_call()` and `_do_api_call_async()` API changes

I made these more plainly wrap the `requests.request()` and `asyncio.ClientSession.request` APIs. It did not seem like there was much reason for these APIs to deviate. Deviations include:

- It had a parameter called `json` that maps to either the `data=json` or `params=json` kwargs in the `requests.request()` call. This seems less desirable than just passing `data=data`, `params=params`, or `json=json`. Specifically, the `json` kwarg is confusing from the perspective of the `requests` and `aiohttp` APIs: in those APIs, `json=` expects a dict and `data=` expects a string, but in the FivetranHook API, `json=` is a string that gets mapped to `data=`.
- The first 2 args were a tuple for method and URL, whereas these could just be args (like `requests.request()`)
- For the internals: it is unnecessarily burdensome from a Mypy perspective, and also inflexible from a future-proofing perspective, to do an if-elif on the method to get the implementation when `requests.request()` can do that for you.

That said, **changing this is probably annoying for users** as there may be many users using this function. So I made sure the old API works and gave a deprecation warning via the following:

```python
if is_container(method) and len(method) == 2:
    import warnings
    warnings.warn(
        "The API for _do_api_call() has changed to closer match the"
        " requests.request() API. Please update your code accordingly.",
        DeprecationWarning,
        stacklevel=2
    )
    method, endpoint = method
elif endpoint is None:
    raise TypeError(
        f"{self.__class__.__name__}._do_api_call() missing 1 required"
        f" positional argument: 'endpoint'"
    )
```

Lastly, I removed the test for `"UNKNOWN"`. The reason why is because `requests.request()` can handle any method passed in, and the Fivetran API itself should be in charge of rejecting bad methods. (Note that this is the only nontrivial test change I made; all other test cases pass with essentially no changes, meaning my code is more or less a refactor.)

**TLDR:**

- `_do_api_call()` now matches `requests.request()`
- I made this backwards compatible with a `DeprecationWarning`

## "Breaking" changes

I put "breaking" in quotes because these are very minor API features that were not being used by anything. That said they still constitute breaking changes, in some sense, if we consider changing the `FivetranHook`'s method return types to constitute breaking changes (even if those return types are not so useful).

- `FivetranHook.check_connector` now returns the `connector_details` dict object. (So that `prep_connector` can avoid making an unnecessary 3rd API call.)
- `FivetranHook.prep_connector` now returns `None` instead of `Literal[True]`.

## Use `hook` as cached property

It is more idiomatic in Airflow code for the `hook` for an operator to be called as a `cached_property`, so I swapped the operator and sensor code to implement this pattern, and added a deprecation warning for `_get_hook()`.

### CODE REVIEW

File: fivetran_provider_async/operators.py [hunk 1]
```diff
@@ -100,8 +101,21 @@ def execute(self, context: Context) -> None | str:
                     ),
                     method_name="execute_complete",
                 )
+            return None
 
     def _get_hook(self) -> FivetranHook:
+        import warnings
+
+        warnings.warn(
```
Comment (pankajastro) [inline fivetran_provider_async/operators.py:109]: _get_hook is private so should be ok to remove it

**Review by pankajastro — COMMENTED** [review]
Look like ci is not happy

File: fivetran_provider_async/hooks.py [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
 from time import sleep
-from typing import Any, Dict, cast
+from typing import TYPE_CHECKING, Any, Dict
 
 import aiohttp
 import pendulum
 import requests
 from aiohttp import ClientResponseError
-from airflow.exceptions import AirflowException
+from airflow.exceptions import AirflowException, AirflowProviderDeprecationWarning
```
[the remark this answers is not in the document]
Comment (dwreeves) [reply fivetran_provider_async/hooks.py:13]: Ah, good point.

We can either set `AirflowProviderDeprecationWarning = DeprecationWarning` on `ImportError`, or alternatively just use base `DeprecationWarning` and leave it at that.

I will go for the simpler approach (option 2).

File: fivetran_provider_async/operators.py (hunk 1 above)
Comment (dwreeves) [reply fivetran_provider_async/operators.py:109]: Done.

File: fivetran_provider_async/hooks.py [hunk 3]
```diff
@@ -63,7 +69,7 @@ def _get_airflow_version() -> str:
         """
         Fetch and return the current Airflow version
         from aws provider
-        https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/base_aws.py#L486
+        https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/base_aws.py#L536
```
[the remark this answers is not in the document]
Comment (dwreeves) [reply fivetran_provider_async/hooks.py:72]: Done.

**Comment by dwreeves** [comment]
CI failed because `isort` moved the `type: ignore[import]` comment... 😅 Fixed!

**Comment by dwreeves** [comment]
No problem, thanks for the merge. I will work on #49 next.
